File: app/src/main/java/mostafa/ma/saleh/gmail/com/editcreditdemo/MainActivity.java
package mostafa.ma.saleh.gmail.com.editcreditdemo;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.RadioGroup;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.IdRes;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import mostafa.ma.saleh.gmail.com.editcredit.EditCredit;

public class MainActivity extends AppCompatActivity implements RadioGroup.OnCheckedChangeListener, View.OnClickListener, CheckBox.OnCheckedChangeListener {

    private EditCredit mEditCredit;
    private RadioGroup separatorsRadioGroup;
    private RadioGroup gravityRadioGroup;
    private Button validateButton;
    private Button getNumberButton;
    private Button getTypeButton;
    private CheckBox visaCheckBox;
    private CheckBox masterCardCheckBox;
    private CheckBox americanExpressCheckBox;
    private CheckBox discoverCheckBox;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViewsById();
        setListeners();
    }

    private void setListeners() {
        separatorsRadioGroup.setOnCheckedChangeListener(this);
        gravityRadioGroup.setOnCheckedChangeListener(this);
        validateButton.setOnClickListener(this);
        getNumberButton.setOnClickListener(this);
        getTypeButton.setOnClickListener(this);
        visaCheckBox.setOnCheckedChangeListener(this);
        masterCardCheckBox.setOnCheckedChangeListener(this);
        americanExpressCheckBox.setOnCheckedChangeListener(this);
        discoverCheckBox.setOnCheckedChangeListener(this);
    }

    private void findViewsById() {
        mEditCredit = findViewById(R.id.editcredit);
        separatorsRadioGroup = findViewById(R.id.separators_radio_group);
        gravityRadioGroup = findViewById(R.id.gravity_radio_group);
        validateButton = findViewById(R.id.btn_validate);
        getNumberButton = findViewById(R.id.btn_get_number);
        getTypeButton = findViewById(R.id.btn_get_type);
        visaCheckBox = findViewById(R.id.chk_visa);
        masterCardCheckBox = findViewById(R.id.chk_mastercard);
        americanExpressCheckBox = findViewById(R.id.chk_amex);
        discoverCheckBox = findViewById(R.id.chk_discover);
    }

    @Override
    public void onCheckedChanged(RadioGroup group, @IdRes int checkedId) {
        switch (group.getId()) {
            case R.id.separators_radio_group:
                onSeparatorsCheckedChanged(checkedId);
                break;
            case R.id.gravity_radio_group:
                onGravityCheckedChanged(checkedId);
                break;
        }
    }

    private void onSeparatorsCheckedChanged(@IdRes int checkedId) {
        switch (checkedId) {
            case R.id.no_separators_radio_button:
                mEditCredit.setSeparator(EditCredit.Separator.NONE);
                break;
            case R.id.spaces_radio_button:
                mEditCredit.setSeparator(EditCredit.Separator.SPACES);
                break;
            case R.id.dashes_radio_button:
                mEditCredit.setSeparator(EditCredit.Separator.DASHES);
                break;
        }
    }

    private void onGravityCheckedChanged(@IdRes int checkedId) {
        switch (checkedId) {
            case R.id.start_radio_button:
                mEditCredit.setDrawableGravity(EditCredit.Gravity.START);
                break;
            case R.id.end_radio_button:
                mEditCredit.setDrawableGravity(EditCredit.Gravity.END);
                break;
            case R.id.left_radio_button:
                mEditCredit.setDrawableGravity(EditCredit.Gravity.LEFT);
                break;
            case R.id.right_radio_button:
                mEditCredit.setDrawableGravity(EditCredit.Gravity.RIGHT);
                break;
        }
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        List<EditCredit.Card> disabledCards = new ArrayList<>();
        if (visaCheckBox.isChecked()) {
            disabledCards.add(EditCredit.Card.VISA);
        }
        if (masterCardCheckBox.isChecked()) {
            disabledCards.add(EditCredit.Card.MASTERCARD);
        }
        if (americanExpressCheckBox.isChecked()) {
            disabledCards.add(EditCredit.Card.AMEX);
        }
        if (discoverCheckBox.isChecked()) {
            disabledCards.add(EditCredit.Card.DISCOVER);
        }
        mEditCredit.setDisabledCards(disabledCards.toArray(new EditCredit.Card[0]));
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_validate:
                Toast.makeText(this, mEditCredit.isCardValid() ? "Valid" : "Not Valid", Toast.LENGTH_SHORT).show();
                break;
            case R.id.btn_get_number:
                Toast.makeText(this, mEditCredit.getTextWithoutSeparator(), Toast.LENGTH_SHORT).show();
                break;
            case R.id.btn_get_type:
                Toast.makeText(this, getCardTypeString(mEditCredit.getCardType()), Toast.LENGTH_SHORT).show();
                break;
        }
    }

    private String getCardTypeString(EditCredit.Card card) {
        @StringRes
        int stringRes;
        switch (card) {
            case VISA:
                stringRes = R.string.visa;
                break;
            case MASTERCARD:
                stringRes = R.string.mastercard;
                break;
            case AMEX:
                stringRes = R.string.american_express;
                break;
            case DISCOVER:
                stringRes = R.string.discover;
                break;
            default:
                stringRes = R.string.unknown;
                break;
        }
        return getString(stringRes);
    }
}

File: app/src/main/java/mostafa/ma/saleh/gmail/com/editcreditdemo/MainActivity.kt
package mostafa.ma.saleh.gmail.com.editcreditdemo

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.CompoundButton
import android.widget.RadioGroup
import android.widget.Toast
import androidx.annotation.IdRes
import androidx.appcompat.app.AppCompatActivity
import mostafa.ma.saleh.gmail.com.editcredit.EditCredit
import mostafa.ma.saleh.gmail.com.editcreditdemo.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity(), RadioGroup.OnCheckedChangeListener,
    CompoundButton.OnCheckedChangeListener {

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(LayoutInflater.from(this))
        setContentView(binding.root)
        addListeners()
    }

    private fun addListeners() {
        binding.separatorsRadioGroup.setOnCheckedChangeListener(this)
        binding.gravityRadioGroup.setOnCheckedChangeListener(this)
        binding.visaCheckBox.setOnCheckedChangeListener(this)
        binding.masterCardCheckBox.setOnCheckedChangeListener(this)
        binding.americanExpressCheckBox.setOnCheckedChangeListener(this)
        binding.discoverCheckBox.setOnCheckedChangeListener(this)
        binding.dinersCheckBox.setOnCheckedChangeListener(this)
    }

    override fun onCheckedChanged(group: RadioGroup, @IdRes checkedId: Int) {
        when (group.id) {
            R.id.separatorsRadioGroup -> onSeparatorsCheckedChanged(checkedId)
            R.id.gravityRadioGroup -> onGravityCheckedChanged(checkedId)
        }
    }

    private fun onSeparatorsCheckedChanged(@IdRes checkedId: Int) {
        binding.editCredit.separator = when (checkedId) {
            R.id.spacesRadioButton -> EditCredit.Separator.SPACES
            R.id.dashesRadioButton -> EditCredit.Separator.DASHES
            else -> EditCredit.Separator.NONE
        }
    }

    private fun onGravityCheckedChanged(@IdRes checkedId: Int) {
        binding.editCredit.drawableGravity = when (checkedId) {
            R.id.startRadioButton -> EditCredit.Gravity.START
            R.id.leftRadioButton -> EditCredit.Gravity.LEFT
            R.id.rightRadioButton -> EditCredit.Gravity.RIGHT
            else -> EditCredit.Gravity.END
        }
    }

    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
        val disabledCards = mutableListOf<EditCredit.Card>()
        if (binding.visaCheckBox.isChecked) {
            disabledCards.add(EditCredit.Card.VISA)
        }
        if (binding.masterCardCheckBox.isChecked) {
            disabledCards.add(EditCredit.Card.MASTERCARD)
        }
        if (binding.americanExpressCheckBox.isChecked) {
            disabledCards.add(EditCredit.Card.AMEX)
        }
        if (binding.discoverCheckBox.isChecked) {
            disabledCards.add(EditCredit.Card.DISCOVER)
        }
        if (binding.dinersCheckBox.isChecked) {
            disabledCards.add(EditCredit.Card.DINERS)
        }
        binding.editCredit.setDisabledCards(*disabledCards.toTypedArray())
    }

    fun onClick(view: View) {
        val message = when (view.id) {
            R.id.validateButton -> if (binding.editCredit.isCardValid) "Valid" else "Not Valid"
            R.id.getNumberButton -> binding.editCredit.textWithoutSeparator
            R.id.getTypeButton -> getCardTypeString(binding.editCredit.cardType)
            else -> return
        }
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    private fun getCardTypeString(card: EditCredit.Card) =
        getString(
            when (card) {
                EditCredit.Card.VISA -> R.string.visa
                EditCredit.Card.MASTERCARD -> R.string.mastercard
                EditCredit.Card.AMEX -> R.string.american_express
                EditCredit.Card.DISCOVER -> R.string.discover
                EditCredit.Card.DINERS -> R.string.diners_club_international
                else -> R.string.unknown
            }
        )
}
